### PR TITLE
[ATO-1652] Add gRPC support

### PR DESCRIPTION
**Pull request by radovanZRasa**
Add support to communicate with Rasa over gRPC protocol.

**Status (please check what you already did)**:
- [x] made PR ready for code review
- [ ] added some tests for the functionality
- [ ] updated the documentation in the [rasaHQ/rasa](https://github.com/rasaHQ/rasa)
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa-sdk/tree/main/changelog) for instructions)
- [ ] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa-sdk#code-style) for instructions)

### CODE REVIEW

**Comment by radovanZRasa** [comment]
Still working on integration tests, unit tests and CI.

File: Makefile [hunk 1]
```diff
@@ -62,3 +61,13 @@ cleanup-generated-changelog:
 release:
 	poetry run python scripts/release.py
 
+download-protoc-compiler:
```
Comment (radovanZRasa) [inline Makefile:64]: This is not needed and will be removed.

File: rasa_sdk/grpc_errors.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
+from __future__ import annotations
+
+from pydantic import BaseModel, Field
+
+from enum import Enum
+
+
+class ActionExecutionFailed(BaseModel):
```
Comment (radovanZRasa) [inline rasa_sdk/grpc_errors.py:8]: These errors are used to provide more context to the receiver into what actually happened.

File: rasa_sdk/grpc_server.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+        request: WebhookRequest,
+        context,
+    ) -> action_webhook_pb2.WebhookResponse:
+        """Handle RPC request for the webhook.
+
+        Args:
+            request: The webhook request.
+            context: The context of the request.
+        """
+        await asyncio.sleep(50)
```
Comment (radovanZRasa) [inline rasa_sdk/grpc_server.py:69]: Test code line. Will be removed.

File: rasa_sdk/interfaces.py [hunk 4]
```diff
@@ -46,7 +46,7 @@ def __init__(
         followup_action: Optional[Text],
         active_loop: Dict[Text, Any],
         latest_action_name: Optional[Text],
-        stack: List[Dict[Text, Any]] = [],
+        stack: Optional[List[Dict[Text, Any]]] = None,
```
Comment (ancalita) [inline rasa_sdk/interfaces.py:49]: Such a good catch!

File: rasa_sdk/tracing/utils.py [hunk 5]
```diff
@@ -17,11 +18,12 @@ def get_tracer_provider(endpoints_file: str) -> Optional[TracerProvider]:
...
 
 def get_tracer_and_context(
-    tracer_provider: Optional[TracerProvider], request: Request
+    tracer_provider: Optional[TracerProvider], request: Union[Request, WebhookRequest]
 ) -> Tuple[Any, Any, Text]:
     """Gets tracer and context."""
     span_name = "create_app.webhook"
-    if tracer_provider is None:
+
+    if tracer_provider is None or isinstance(request, WebhookRequest):
```
Comment (ancalita) [inline rasa_sdk/tracing/utils.py:26]: Yeah we will need to update the code on rasa-private side too to inject the trace context in the grpc request. I am searching for some docs / looking in opentelemetry to see if there's any guidance.

Comment (ancalita) [reply rasa_sdk/tracing/utils.py:26]: Couldn't find any guidance I'm afraid :(

File: rasa_sdk/interfaces.py (hunk 4 above)
Comment (radovanZRasa) [reply rasa_sdk/interfaces.py:49]: Thanks!

File: rasa_sdk/tracing/utils.py (hunk 5 above)
Comment (radovanZRasa) [reply rasa_sdk/tracing/utils.py:26]: I'll try to look it up.

File: Makefile [hunk 6]
```diff
@@ -37,7 +36,7 @@ formatter:
 
 lint:
 	poetry run ruff check rasa_sdk tests --ignore D
-	poetry run black --check rasa_sdk tests
+	poetry run black --exclude="rasa_sdk/grpc_py" --check rasa_sdk tests
```
Comment (ancalita) [inline Makefile:39]: Could we add this to the pyproject config instead?

File: Makefile [hunk 7]
```diff
@@ -62,3 +61,11 @@ cleanup-generated-changelog:
...
 	poetry run python scripts/release.py
 
+generate-grpc:
+	 python -m grpc_tools.protoc \
+	 	-Irasa_sdk/grpc_py=./proto \
+	 	--python_out=. \
+	 	--grpc_python_out=. \
+	 	--pyi_out=. \
+	 	proto/action_webhook.proto \
+	 	proto/health.proto
```
Comment (ancalita) [inline Makefile:71]: Missing new line.

File: proto/health.proto [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+
+package grpc.health.v1;
+
+message HealthCheckRequest {}
+
+message HealthCheckResponse {}
+
+service HealthService {
+  rpc Check(HealthCheckRequest) returns (HealthCheckResponse);
+}
```
Comment (ancalita) [inline proto/health.proto:11]: Missing new line

File: rasa_sdk/__main__.py [hunk 9]
```diff
@@ -18,20 +22,32 @@ def main_from_args(args):
...
+        )
+    else:
+        run(
+            args.actions,
+            args.port,
+            args.cors,
+            args.ssl_certificate,
+            args.ssl_keyfile,
+            args.ssl_password,
+            args.auto_reload,
```
Comment (ancalita) [inline rasa_sdk/__main__.py:44]: I see this is missing from run_grpc, do we need to issue a warning in case this gets set when using grpc?

Comment (ancalita) [reply rasa_sdk/__main__.py:44]: Or should we make this feature available with grpc too?

File: rasa_sdk/grpc_server.py [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+        """Handle RPC request for the health check.
+
+        Args:
+            request: The health check request.
+            context: The context of the request.
+
+        Returns:
+            gRPC response.
+        """
+        response = HealthCheckResponse()
```
Comment (ancalita) [inline rasa_sdk/grpc_server.py:72]: The name of this method implies an action is executed, however we're not using any of the method args, am I missing something?

File: rasa_sdk/grpc_server.py [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+        """Initializes the ActionServerWebhook.
+
+        Args:
+            tracer_provider: The tracer provider.
+            executor: The action executor.
+        """
+        self.tracer_provider = tracer_provider
+        self.executor = executor
+
+    async def Actions(self, request: ActionsRequest, context) -> ActionsResponse:
```
Comment (ancalita) [inline rasa_sdk/grpc_server.py:93]: Same question here about not using the method args as above. Also context is missing type hint.

**Review by ancalita — COMMENTED** [review]
Left a few minor comments, are you planning to tackle tracing in a future PR?

File: Makefile (hunk 6 above)
Comment (radovanZRasa) [reply Makefile:39]: I tried, it does not work.

File: rasa_sdk/grpc_server.py (hunk 10 above)
Comment (radovanZRasa) [reply rasa_sdk/grpc_server.py:72]: It was the cleanest way to provide health check when action server is run in grpc model. We do same no-op in http as well.

File: rasa_sdk/grpc_server.py (hunk 11 above)
Comment (radovanZRasa) [reply rasa_sdk/grpc_server.py:93]: I tried but I could not find the type of the context anywhere.

Considering method args, this is the cleanest way to do it according to [gRPC docs](https://protobuf.dev/programming-guides/api/#unique-protos). 
It provides good options for backwards compatibility later on.

Comment (radovanZRasa) [reply rasa_sdk/grpc_server.py:96]: Found the type of context after some digging through docs. It's `grpc.aio.ServicerContext`.

File: rasa_sdk/tracing/utils.py (hunk 5 above)
Comment (radovanZRasa) [reply rasa_sdk/tracing/utils.py]: @ancalita I added tracing for `GRPCActionServerWebhook.Webhook`.

File: rasa_sdk/grpc_server.py (hunk 11 above)
Comment (ancalita) [reply rasa_sdk/grpc_server.py:97]: Nice 💯 could we add it in?

File: rasa_sdk/tracing/utils.py [hunk 12]
```diff
@@ -17,10 +17,11 @@ def get_tracer_provider(endpoints_file: str) -> Optional[TracerProvider]:
 
 
 def get_tracer_and_context(
-    tracer_provider: Optional[TracerProvider], request: Request
+    tracer_provider: Optional[TracerProvider], request: Union[Request]
```
Comment (ancalita) [inline rasa_sdk/tracing/utils.py:20]: This had the WebhookRequest before I think in the typing hint, could you please re-add it?

**Review by ancalita — APPROVED** [review]
Awesome 💯 🚀

File: rasa_sdk/tracing/utils.py (hunk 12 above)
Comment (ancalita) [reply rasa_sdk/tracing/utils.py:20]: But that serves a different purpose, we need this for passing the rasa-private context to the sdk 🤔

Comment (ancalita) [reply rasa_sdk/tracing/utils.py:20]: The tracer in the webhook instruments the webhook separately.

**Pull request merged**